geps/gep-1897.md

```diff
@@ -125,7 +125,231 @@ This GEP is the outcome of the TLS use cases #4 and #5 in
+- Version: specifies the minimum TLS version that the connection may use (default TLSv1.2)
+- Visibility: the visibility level could be all or none, indicating that if the connection failed due to validation
+failures, it would drop the connection silently if the visibility level were none, and report an error if the visibility
+level were all.  (default to all)
+
+```go
+// TLSConnectionPolicy provides a way to publish TLS configuration
+// that enables a gateway client to connect to a backend pod.
+type TLSConnectionPolicy struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec defines the desired state of TLSConnectionPolicy.
+    Spec TLSConnectionPolicySpec `json:"spec"`
+
+    // Status defines the current state of TLSConnectionPolicy.
+    Status TLSConnectionPolicyStatus `json:"status,omitempty"`
+}
+
+// TLSConnectionPolicySpec defines the desired state of
+// TLSConnectionPolicy.
+// Note: there is no Override or Default policy configuration.
+type TLSConnectionPolicySpec struct {
+    // TargetRef identifies an API object to apply policy to.
+    TargetRef gatewayv1a2.PolicyTargetReference `json:"targetRef"`
+
+    // TLS contains TLS connection policy configuration.
+    TLS *TLSConnectionPolicyConfig `json:”tls”`
+}
+
+// TLSConnectionPolicyConfig contains TLS connection policy configuration.
+type TLSConnectionPolicyConfig struct {
+    // TrustedCACertRefs contains one or more references to
+    // Kubernetes objects that contain TLS certificates, which are
+    // used to establish a TLS handshake between the gateway and
+    // backend pod.
+    //
+    // A single TrustedCACertRef to a Kubernetes Secret has "Core"
+    // support.  Implementations MAY choose to support attaching
+    // multiple certificates to a backend, but this behavior is 
+    // implementation-specific.
+    //
+    // References to a resource in a different namespace are 
+    // invalid.
+    // 
+    // This field is required for any TLSConnectionPolicyConfig.
+    // 
+    // Support: Core - A single reference to a Kubernetes Secret of type kubernetes.io/tls
+    //
+    // Support: Implementation-specific (More than one reference or other resource types)
+    //
+    // +kubebuilder:validation:MaxItems=64
+    // +kubebuilder:validation:MinItems=1
+    TrustedCACertRefs []SecretObjectReference `json:”trustedCACertRefs”`
+
+    // Port is the network port that the implementation watches to
+    // know if the connection should be TLS and the targetRef’s
+    // certificate should be validated by the certs in TrustedCACertRefs
+    // If empty, then all ports for the targetRef are watched to know
+    // if the connection should be TLS, the targetRef’s certificate
+    // should be validated by the certs in TrustedCACertRefs, and a
+    // status delivered in the response for validation failures.
+    Port PortNumber `json:port,omitempty`
```

Comment:
So the target of this policy is a `Service` and as far as I know the `Service` will have a `targetPort` (or defaulted to service port?). Can we just use that instead so this port is not needed? Otherwise how does the implementation work when the `targetPort` in the `Service` and the `Port` here are different?

Reply:
In Gateway API, a port reference is traditionally part of a Service reference (BackendRef is best example) and it refers to the Port the Service is listening on, not the targetPort (port the Service backends are listening on). The reason this is needed is that it's fairly common for Kubernetes Services to listen on multiple ports and protocols.

So given the following example, this would be referring to port 80, and specifically backendRefs that are targeting this Service on port 80.

```
apiVersion: v1
kind: Service
metadata:
  name: my-service
spec:
  selector:
    app.kubernetes.io/name: MyApp
  ports:
    - protocol: TCP
      port: 80
      targetPort: 9376
```

Reply:
> ...it's fairly common for Kubernetes Services to listen on multiple ports and protocols.

Got it. So in the case where a Kubernetes service listens on a single port, the `Port` field here (on line 243) has to match the service port. Put another way, the port can be inferred in such a case. Just confirming my understanding